### PR TITLE
pool: allow updating deviceClass on existing pool

### LINKED ISSUE #12714: Support changing the device class on a pool
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
If the deviceClass is updated on a pool (in a CephBlockPool, CephFilesystem, or CephObjectStore), the CRUSH rule needs to be re-created with the new device class. 

**What is use case behind this feature:**
The device class is a part of the CRUSH rule rook generates for the pool. After the initial reconcile, the CRUSH rule is currently never updated by Rook. If the deviceClass changes, it is necessary to re-create the rule so that Ceph will move the PGs according to the new failure domain. 

In past discussions this has been considered not supported, only to be done manually from the toolbox, or else to be implemented with the kubectl plugin. However, it seems safe enough of an operation from Ceph's perspective that it can be better automated in the pool reconcile in Rook. Then there is no need for manual or kubectl plugin commands.

### PR DESCRIPTION
**Pull request by subhamkrai**
**Description of your changes:**
This commits add check to update the deviceClass on existing pool.

**Which issue is resolved by this Pull Request:**
Resolves #12714 #12263 

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Review by subhamkrai — COMMENTED** [review]
still needs testing but meantime feel free to review

File: pkg/daemon/ceph/client/pool.go [hunk 1]
```diff
@@ -690,6 +689,10 @@ func createReplicationCrushRule(context *clusterd.Context, clusterInfo *ClusterI
 
 	args := []string{"osd", "crush", "rule", "create-replicated", ruleName, crushRoot, failureDomain}
 
+	if pool.DeviceClass != "" {
+		args = append(args, pool.DeviceClass)
```
[the remark this answers is not in the document]
Comment (subhamkrai) [reply pkg/daemon/ceph/client/pool.go]: yes, it was added by mistake

File: pkg/daemon/ceph/client/pool.go [hunk 2]
```diff
@@ -471,7 +471,7 @@ func createReplicatedPoolForApp(context *clusterd.Context, clusterInfo *ClusterI
 
 	logger.Infof("reconciling replicated pool %s succeeded", pool.Name)
 
-	if checkFailureDomain {
+	if checkFailureDomain || pool.PoolSpec.DeviceClass != "" {
```
[the remark this answers is not in the document]
Comment (subhamkrai) [reply pkg/daemon/ceph/client/pool.go:474]: we don't need to check similar to https://github.com/rook/rook/blob/72452f0558281c0b57f93017c481f5518ca12b86/pkg/daemon/ceph/client/pool.go#L483-L486
since deviceClass can be empty or not always specified

File: pkg/daemon/ceph/client/pool.go [hunk 3]
```diff
@@ -498,8 +498,7 @@ func ensureFailureDomain(context *clusterd.Context, clusterInfo *ClusterInfo, cl
 	}
 	currentFailureDomain := extractFailureDomain(rule)
 	if currentFailureDomain == pool.FailureDomain {
-		logger.Debugf("pool %q has the expected failure domain %q", pool.Name, pool.FailureDomain)
-		return nil
+		logger.Info("pool %q has the expected failure domain %q", pool.Name, pool.FailureDomain)
```
[the remark this answers is not in the document]
Comment (subhamkrai) [reply pkg/daemon/ceph/client/pool.go]: right added the check for this

**Comment by subhamkrai** [comment]
@travisn how we can verify the changes?

**Comment by travisn** [comment]
> @travisn how we can verify the changes?

1. Create a cluster 
2. Create a pool with no `deviceClass` set
3. Update the pool to set the `deviceClass` to the same as one of the OSDs
4. Verify the pool's crush rule was created as expected

Another scenario could be to start with one device class, and then update it to a different device class.

**Comment by subhamkrai** [comment]
> > @travisn how we can verify the changes?
> 
> 1. Create a cluster
> 2. Create a pool with no `deviceClass` set
> 3. Update the pool to set the `deviceClass` to the same as one of the OSDs
> 4. Verify the pool's crush rule was created as expected
> 
> Another scenario could be to start with one device class, and then update it to a different device class.

I tested both the cases and seems like it is working,

```
pool 2 'subham' replicated size 1 min_size 1 crush_rule 4 object_hash rjenkins pg_num 32 pgp_num 32 autoscale_mode on last_change 39 lfor 0/0/31 flags hashpspool,selfmanaged_snaps stripe_width 0 application rbd read_balance_score 1.00
```

```
 {
        "rule_id": 4,
        "rule_name": "subham_osd_hdd",
        "type": 1,
        "steps": [
            {
                "op": "take",
                "item": -2,
                "item_name": "default~hdd"
            },
            {
                "op": "choose_firstn",
                "num": 0,
                "type": "osd"
            },
            {
                "op": "emit"
            }
        ]
    }
```

**Review by subhamkrai — COMMENTED** [review]
@travisn do you think we should update the function names and logging as it mostly talks about updating failureDomain or related failureDomain and not about deviceClass?

File: pkg/daemon/ceph/client/pool.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (travisn) [reply pkg/daemon/ceph/client/pool.go:474]: If failureDomain is blank, but the deviceClass is changed, how do we get past the return on line 485?

**Review by travisn — COMMENTED** [review]
> @travisn do you think we should update the function names and logging as it mostly talks about updating failureDomain or related failureDomain and not about deviceClass?

Yes, sounds good to update the function names to be more generic.

File: pkg/daemon/ceph/client/pool.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (subhamkrai) [reply pkg/daemon/ceph/client/pool.go:474]: I missed that since in the pool example failure domain was set, now I have added to check for deviceClass also.

File: pkg/daemon/ceph/client/pool.go [hunk 4]
```diff
@@ -496,17 +496,24 @@ func ensureFailureDomain(context *clusterd.Context, clusterInfo *ClusterInfo, cl
...
+	currentFailureDomain, currentDeviceClass := extractPoolDetails(rule)
+	if currentFailureDomain == pool.FailureDomain && currentDeviceClass == pool.DeviceClass {
+		logger.Info("pool %q has the expected failure domain %q and deviceClass", pool.Name, pool.FailureDomain, pool.DeviceClass)
 		return nil
 	}
+
 	if currentFailureDomain == "" {
 		logger.Warningf("failure domain not found for crush rule %q, proceeding to create a new crush rule", details.CrushRule)
 	}
+	if currentFailureDomain == "" {
```
Comment (travisn) [inline pkg/daemon/ceph/client/pool.go:508]:
```suggestion
	if currentDeviceClass != pool.DeviceClass {
```

File: pkg/daemon/ceph/client/pool.go [hunk 5]
```diff
@@ -496,17 +496,24 @@ func ensureFailureDomain(context *clusterd.Context, clusterInfo *ClusterInfo, cl
...
-	currentFailureDomain := extractFailureDomain(rule)
-	if currentFailureDomain == pool.FailureDomain {
-		logger.Debugf("pool %q has the expected failure domain %q", pool.Name, pool.FailureDomain)
+	currentFailureDomain, currentDeviceClass := extractPoolDetails(rule)
+	if currentFailureDomain == pool.FailureDomain && currentDeviceClass == pool.DeviceClass {
+		logger.Info("pool %q has the expected failure domain %q and deviceClass", pool.Name, pool.FailureDomain, pool.DeviceClass)
 		return nil
 	}
+
 	if currentFailureDomain == "" {
```
Comment (travisn) [inline pkg/daemon/ceph/client/pool.go:505]:
```suggestion
	if currentFailureDomain != pool.FailureDomain {
```

File: pkg/daemon/ceph/client/pool.go [hunk 6]
```diff
@@ -496,17 +496,24 @@ func ensureFailureDomain(context *clusterd.Context, clusterInfo *ClusterInfo, cl
...
 		logger.Warningf("failure domain not found for crush rule %q, proceeding to create a new crush rule", details.CrushRule)
 	}
+	if currentFailureDomain == "" {
+		logger.Warningf("deviceClass not found for crush rule %q, proceeding to create a new crush rule", details.CrushRule)
```
Comment (travisn) [inline pkg/daemon/ceph/client/pool.go:509]: Let's change this one and on line 506 to an info message
```suggestion
		logger.Infof("deviceClass not found for crush rule %q, proceeding to create a new crush rule", details.CrushRule)
```

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/daemon/ceph/client/pool.go (hunk 4 above)
Comment (subhamkrai) [reply pkg/daemon/ceph/client/pool.go]: make sense

File: pkg/daemon/ceph/client/pool.go [hunk 7]
```diff
@@ -496,17 +496,24 @@ func ensureFailureDomain(context *clusterd.Context, clusterInfo *ClusterInfo, cl
 	if err != nil {
 		return errors.Wrapf(err, "failed to get crush rule %q", details.CrushRule)
 	}
-	currentFailureDomain := extractFailureDomain(rule)
-	if currentFailureDomain == pool.FailureDomain {
-		logger.Debugf("pool %q has the expected failure domain %q", pool.Name, pool.FailureDomain)
+	currentFailureDomain, currentDeviceClass := extractPoolDetails(rule)
+	if currentFailureDomain == pool.FailureDomain && currentDeviceClass == pool.DeviceClass {
+		logger.Info("pool %q has the expected failure domain %q and deviceClass", pool.Name, pool.FailureDomain, pool.DeviceClass)
```
Comment (travisn) [inline pkg/daemon/ceph/client/pool.go:501]:
```suggestion
		logger.Info("pool %q has the expected failure domain %q and deviceClass %q", pool.Name, pool.FailureDomain, pool.DeviceClass)
```

**Review by travisn — COMMENTED** [review]
One small issue

File: pkg/daemon/ceph/client/pool.go [hunk 8]
```diff
@@ -471,21 +471,21 @@ func createReplicatedPoolForApp(context *clusterd.Context, clusterInfo *ClusterI
...
 	}
 	return nil
 }
 
-func ensureFailureDomain(context *clusterd.Context, clusterInfo *ClusterInfo, clusterSpec *cephv1.ClusterSpec, pool cephv1.NamedPoolSpec) error {
-	if pool.FailureDomain == "" {
-		logger.Debugf("skipping check for failure domain on pool %q as it is not specified", pool.Name)
+func updatePoolCrushRule(context *clusterd.Context, clusterInfo *ClusterInfo, clusterSpec *cephv1.ClusterSpec, pool cephv1.NamedPoolSpec) error {
+	if pool.FailureDomain == "" && pool.DeviceClass == "" {
+		logger.Debugf("skipping check for failure domain %q and deviceClass on pool %q as it is not specified", pool.Name, pool.DeviceClass)
```
Comment (travisn) [inline pkg/daemon/ceph/client/pool.go:484]: deviceClass is always empty here, no need to format it in the message
```suggestion
		logger.Debugf("skipping check for failure domain and deviceClass on pool %q as it is not specified", pool.Name)
```

File: pkg/daemon/ceph/client/pool.go [hunk 9]
```diff
@@ -496,17 +496,24 @@ func ensureFailureDomain(context *clusterd.Context, clusterInfo *ClusterInfo, cl
 	if err != nil {
 		return errors.Wrapf(err, "failed to get crush rule %q", details.CrushRule)
 	}
-	currentFailureDomain := extractFailureDomain(rule)
-	if currentFailureDomain == pool.FailureDomain {
-		logger.Debugf("pool %q has the expected failure domain %q", pool.Name, pool.FailureDomain)
+	currentFailureDomain, currentDeviceClass := extractPoolDetails(rule)
+	if currentFailureDomain == pool.FailureDomain && currentDeviceClass == pool.DeviceClass {
+		logger.Info("pool %q has the expected failure domain %q and deviceClass %q", pool.Name, pool.FailureDomain, pool.DeviceClass)
```
Comment (travisn) [inline pkg/daemon/ceph/client/pool.go:501]:
```suggestion
		logger.Infof("pool %q has the expected failure domain %q and deviceClass %q", pool.Name, pool.FailureDomain, pool.DeviceClass)
```

**Review by travisn — CHANGES_REQUESTED** [review]
Might need to test a few scenarios and look at the operator log to verify the messages are all formatted properly.

**Review by sp98 — COMMENTED** [review]
@subhamkrai Maybe I'm missing something here, but at what point is the deviceClass of the pool updated in the code changes?

**Comment by subhamkrai** [comment]
> @subhamkrai Maybe I'm missing something here, but at what point is the deviceClass of the pool updated in the code changes?

The code was reading and reconciling the deviceClass pool but it was never applied to the ceph pool as it required to creation new Crush rule based on the deviceClass. This is what the pr is all about

File: pkg/daemon/ceph/client/pool.go (hunk 8 above)
Comment (subhamkrai) [reply pkg/daemon/ceph/client/pool.go:484]: output with testing 
```
 skipping check for failure domain and deviceClass "replicapool" on pool as it is not specified
```

File: pkg/daemon/ceph/client/pool.go [hunk 10]
```diff
@@ -471,21 +471,21 @@ func createReplicatedPoolForApp(context *clusterd.Context, clusterInfo *ClusterI
...
 	}
 	return nil
 }
 
-func ensureFailureDomain(context *clusterd.Context, clusterInfo *ClusterInfo, clusterSpec *cephv1.ClusterSpec, pool cephv1.NamedPoolSpec) error {
-	if pool.FailureDomain == "" {
-		logger.Debugf("skipping check for failure domain on pool %q as it is not specified", pool.Name)
+func updatePoolCrushRule(context *clusterd.Context, clusterInfo *ClusterInfo, clusterSpec *cephv1.ClusterSpec, pool cephv1.NamedPoolSpec) error {
+	if pool.FailureDomain == "" && pool.DeviceClass == "" {
+		logger.Debugf("skipping check for failure domain and deviceClass %q on pool as it is not specified", pool.Name)
```
Comment (travisn) [inline pkg/daemon/ceph/client/pool.go:484]:
```suggestion
		logger.Debugf("skipping check for failure domain and deviceClass on pool %q as it is not specified", pool.Name)
```

File: pkg/daemon/ceph/client/pool.go [hunk 11]
```diff
@@ -496,17 +496,24 @@ func ensureFailureDomain(context *clusterd.Context, clusterInfo *ClusterInfo, cl
...
+	currentFailureDomain, currentDeviceClass := extractPoolDetails(rule)
+	if currentFailureDomain == pool.FailureDomain && currentDeviceClass == pool.DeviceClass {
+		logger.Infof("pool %q has the expected failure domain %q and deviceClass %q", pool.Name, pool.FailureDomain, pool.DeviceClass)
 		return nil
 	}
-	if currentFailureDomain == "" {
-		logger.Warningf("failure domain not found for crush rule %q, proceeding to create a new crush rule", details.CrushRule)
+
+	if currentFailureDomain != pool.FailureDomain {
+		logger.Infof("failure domain not found for crush rule %q, proceeding to create a new crush rule", details.CrushRule)
```
Comment (travisn) [inline pkg/daemon/ceph/client/pool.go:506]: How about this wording?
```suggestion
		logger.Infof("creating a new crush rule for changed failure domain on crush rule %q", details.CrushRule)
```

File: pkg/daemon/ceph/client/pool.go [hunk 12]
```diff
@@ -496,17 +496,24 @@ func ensureFailureDomain(context *clusterd.Context, clusterInfo *ClusterInfo, cl
...
+	}
+	if currentDeviceClass != pool.DeviceClass {
+		logger.Infof("deviceClass not found for crush rule %q, proceeding to create a new crush rule", details.CrushRule)
```
Comment (travisn) [inline pkg/daemon/ceph/client/pool.go:509]:
```suggestion
		logger.Infof("creating a new crush rule for changed deviceClass on crush rule %q", details.CrushRule)
```

**Review by travisn — CHANGES_REQUESTED** [review]
Just some nits on the log messages. Also, were you able to test this manually since the integration tests don't change these settings?

**Comment by subhamkrai** [comment]
> > > @travisn how we can verify the changes?
> > 
> > 
> > 
> > 1. Create a cluster
> > 2. Create a pool with no `deviceClass` set
> > 3. Update the pool to set the `deviceClass` to the same as one of the OSDs
> > 4. Verify the pool's crush rule was created as expected
> > 
> > Another scenario could be to start with one device class, and then update it to a different device class.
> 
> I tested both the cases and seems like it is working,
> 
> ```
> pool 2 'subham' replicated size 1 min_size 1 crush_rule 4 object_hash rjenkins pg_num 32 pgp_num 32 autoscale_mode on last_change 39 lfor 0/0/31 flags hashpspool,selfmanaged_snaps stripe_width 0 application rbd read_balance_score 1.00
> ```
> 
> ```
>  {
>         "rule_id": 4,
>         "rule_name": "subham_osd_hdd",
>         "type": 1,
>         "steps": [
>             {
>                 "op": "take",
>                 "item": -2,
>                 "item_name": "default~hdd"
>             },
>             {
>                 "op": "choose_firstn",
>                 "num": 0,
>                 "type": "osd"
>             },
>             {
>                 "op": "emit"
>             }
>         ]
>     }
> ```

yes, I did tested manually above is the output

**Review by travisn — APPROVED** [review]

**Comment by anthonyeleven** [comment]
I think this has the potential for disruption, and I'd like to add a warning to the docs.  To that end, I'd like to understand under what conditions this will be applied.  I have pools where I had to change the CRUSH rule manually, and a pool I created manually because Rook couldn't do it.  I want to be sure to understand if/when Rook could unilaterally do something impactful here, and document it before it bites others (or me)


https://rook-io.slack.com/archives/C741SA742/p1700244562501209

**Comment by travisn** [comment]
If the `deviceClass` and/or the `failureDomain` are set in the pool custom resource, Rook will ensure the crush rule has those matching properties. If those properties are both blank, Rook will not attempt to update the crush rule as seen [here](https://github.com/rook/rook/blob/master/pkg/daemon/ceph/client/pool.go#L483).
